### ISSUE TITLE
What's the expected input ?

### ISSUE THREAD

**Issue by LAUBENicolas** [qubvel/segmentation_models#395]
Hello, 
I would like to know if a particular input is expected and what type of preprocessing was used for the original training ?

**Comment by JordanMakesMaps** [comment]
You should use the `preprocess_input()` as described in the readme for whichever backbone you are using. It will use the same preprocessing method as was used when that encoder was originally trained on the ImageNet dataset.

**Comment by LAUBENicolas** [comment]
Thank you for your help @JordanMakesMaps. I'm doing binary segmentation. If I understood well, the dimension of the model the pretrained model resnet on imagenet is  (nb images, images_width, images_height, 3 channels) for the images with values ranging from 0 to 255. and (nb images, images_width, images_height, **1**) for the masks with values in {0, 1}.

**Comment by JordanMakesMaps** [comment]
Yes, for ResNet the range of image values is from [0, 255]; for binary classification the masks should be integers from [0, 1].

For the input dimensions, it depends on what you provide to the model when initializing it.

**Comment by LAUBENicolas** [comment]
Ok, thank you, I didn't change anything in the model definition:
`

        self.model = sm.Unet(backbone_name=BACKBONE_NAME, classes=1, encoder_weights="imagenet", encoder_freeze=True)
        train_dataset, val_dataset = dataset.split_train_test()
        train_dataloader = KerasDataLoader(train_dataset, batch_size=train_batch_size, shuffle=train_shuffle)
        val_dataloader = KerasDataLoader(val_dataset, batch_size=val_batch_size, shuffle=val_shuffle)
        assert train_dataloader[0][0].shape == (train_batch_size, 512, 512, 3)
        assert train_dataloader[0][1].shape == (train_batch_size, 512, 512, 1)
        assert val_dataloader[0][0].shape == (train_batch_size, 512, 512, 3)
        assert val_dataloader[0][1].shape == (train_batch_size, 512, 512, 1)

        # callback functions
        callbacks = [
            keras.callbacks.ModelCheckpoint('./weights/segmentation_UNET.h5', save_weights_only=True, save_best_only=True, mode='min'),
            keras.callbacks.ReduceLROnPlateau(),
        ]

        # compile model
        optimizer = Adam(learning_rate)
        dice_loss = sm.losses.DiceLoss()
        focal_loss = sm.losses.BinaryFocalLoss(alpha=0.75)
        # since nb classes = 1 else use CategorialFocalLoss()
        total_loss = dice_loss + (1 * focal_loss) #dice_loss #+ (1 *
        metrics = [sm.metrics.IOUScore(threshold=0.5), sm.metrics.FScore(threshold=0.5)]
        self.model.compile(optimizer, total_loss, metrics)
`
Should the val_datasetloader be of size (train_batch_size, 512, 512) instead ?
Do you know the size of the input images used in unet ? Here my images are of size (512, 512).

**Comment by LAUBENicolas** [comment]
I think that the input size (512, 512) should be probably changed. 
After training my model, I get a pattern of masks (the same mask is repeated 9 times on my initial image if I overlay predicted mask and initial image).

**Comment by JordanMakesMaps** [comment]
The original paper is [here](https://arxiv.org/abs/1505.04597); you don't need to use the same image dimensions as them, as this repo only recreates UNet in idea, not the exact architecture used in the paper. Nor does it use the weights that were used in that paper.

Did this work for your binary classification task?
```python

    assert train_dataloader[0][0].shape == (train_batch_size, 512, 512, 3)
    assert train_dataloader[0][1].shape == (train_batch_size, 512, 512, 1) 
    assert val_dataloader[0][0].shape == (train_batch_size, 512, 512, 3)
    assert val_dataloader[0][1].shape == (train_batch_size, 512, 512, 1) 
```

**Comment by LAUBENicolas** [comment]
Yes all assertions worked.

I'm using the pretrained "ImageNet" weights for ResNet34.
![image](https://user-images.githubusercontent.com/51950336/94698803-f302d380-0339-11eb-8a21-12bca79fd475.png)
*original image comes from the NSCLC Lungs 4D dataset (Hugo, Geoffrey D., Weiss, Elisabeth, Sleeman, William C., Balik, Salim, Keall, Paul J., Lu, Jun, & Williamson, Jeffrey F. (2016). Data from 4D Lung Imaging of NSCLC Patients. The Cancer Imaging Archive. http://doi.org/10.7937/K9/TCIA.2016.ELN8YGLE)


Here is what I obtain after a few epochs. It is strange that I have a repeated pattern in my predictions.
I froze the encoder during training and a learning rate of ```0.0008```. The dice coeff and val dice are quite low (```0.05``` after 5-6 epochs).

**Comment by JordanMakesMaps** [comment]
Are you concatenating images together at any point before passing the image to the model for prediction? Or concatenating the predicted mask after it is produced? What's the dimensions of the input image and the output mask?

**Comment by LAUBENicolas** [comment]
Yes, I am concatenating images before putting them into the model.
My input image is 2d with only one channel (values between -1000 and 3064).
Th preprocessing step consists of doing a normalization between 0 and 255.0.
Then, I do 
```
images = np.array([images, images, images])
images = images.reshape((images.shape[1], images.shape[2], images.shape[3], images.shape[0]))
preprocess_input = sm.get_preprocessing(DEFAULT_BACKBONE)
images = preprocess_input(images)
```
I'm not using the trick of the initial convolutionnal layer...

**Comment by JordanMakesMaps** [comment]
@LAUBENicolas I think you're concatenating `images` together along the wrong axis and creating a paneled image, which I assume is not intentional:

```python
from skimage import io, color
import numpy as np

images = io.imread("github.png")
images = np.array([images, images, images])
images = images.reshape((images.shape[1], images.shape[2], images.shape[3], images.shape[0]))
io.imshow(images[:, :, 0, :]) # shape = (H, W, ?, ?)
```
![issue](https://user-images.githubusercontent.com/8714759/94825643-0494bb80-03cc-11eb-9ae1-126a55770a7e.png)

Instead, do something like this:

```python
images = color.rgb2gray(io.imread("github.png")) #  a gray scale image for your example
images = np.stack([images] * 3, axis = 2) # shape = (H, W, 3)
io.imshow(images)
```
![fixed](https://user-images.githubusercontent.com/8714759/94826315-cea40700-03cc-11eb-95c6-ff6a575fe09b.png)

I'm also not sure if you actually need to stack grayscale images together to create a false RGB image? You might be able to get away with just changing the input dimensions to the model to something like `(None, None, 1)`? Lemme know how it goes.

**Comment by LAUBENicolas** [comment]
Thank's a lot for your help !
I used the convolutionnal layer instead of the concatenation of images and it solved the pb.
I'll try out your solution also to compare both.

I think that it is mandatory to give a 3 channels input when using the models pretrained on imageNet no ?

**Comment by JordanMakesMaps** [comment]
I think you're right, but you could do a convolutional layer that outputs three filter maps with the same dimensions. Either/or the problem gets solved. Good luck!
